hudi-common/src/main/java/org/apache/hudi/metadata/HoodieTableMetadataUtil.java

```diff
@@ -2808,49 +2812,78 @@ public static HoodieData<HoodieRecord> convertMetadataToPartitionStatRecords(Hoo
+      // TODO(yihua): explore this approach vs RDD based data partitions and fetching files in partition from table metadata
+      //  and construct FSV per partition (this alleviates driver memory at the cost of increasing FS calls to MDT?)
+      SpillableMapBasedFileSystemView fileSystemView = new SpillableMapBasedFileSystemView(
+          tableMetadata, dataMetaClient, dataMetaClient.getActiveTimeline(),
+          consolidatedPathInfos, FileSystemViewStorageConfig.newBuilder().fromProperties(metadataConfig.getProps()).build(),
+          HoodieCommonConfig.newBuilder().fromProperties(metadataConfig.getProps()).build());
+      Map<String, List<FileSlice>> consolidatedFileSliceMap = partitionedWriteStats.keySet().stream()
+          .collect(Collectors.toMap(
+              e -> e,
+              partition -> fileSystemView.getLatestMergedFileSlicesBeforeOrOn(partition, instantTime).collect(toList())));
+
+      HoodiePairData<String, List<HoodieColumnRangeMetadata<Comparable>>> columnRangeMetadata =
+          engineContext.parallelize(partitionedWriteStats.entrySet().stream().collect(Collectors.toList()), parallelism).mapToPair(entry -> {
+            List<HoodieWriteStat> partitionedWriteStat = entry.getValue();
+            final String partitionName = partitionedWriteStat.get(0).getPartitionPath();
+            // Step 1: Collect Column Metadata for Each File part of current commit metadata
+            List<HoodieColumnRangeMetadata<Comparable>> fileColumnMetadata = partitionedWriteStat.stream()
+                .flatMap(writeStat -> translateWriteStatToFileStats(writeStat, dataMetaClient, colsToIndex, partitionStatsIndexVersion).stream()).collect(toList());
+            Set<String> filesWithColumnStats = partitionedWriteStat.stream().map(stat -> stat.getPath()).collect(Collectors.toSet());
+
+            if (shouldScanColStatsForTightBound) {
+              checkState(tableMetadata != null, "tableMetadata should not be null when scanning metadata table");
+              // Collect Column Metadata for Each File part of active file system view of latest snapshot
+              // Get all file names, including log files, in a set from the file slices
+              Set<String> fileNames = consolidatedFileSliceMap.get(partitionName).stream()
+                  .flatMap(fileSlice -> Stream.concat(
+                      Stream.of(fileSlice.getBaseFile().map(HoodieBaseFile::getFileName).orElse(null)),
+                      fileSlice.getLogFiles().map(HoodieLogFile::getFileName)))
+                  .filter(e -> Objects.nonNull(e) && !filesWithColumnStats.contains(e))
+                  .collect(Collectors.toSet());
+              // Fetch metadata table COLUMN_STATS partition records for above files
+              List<HoodieColumnRangeMetadata<Comparable>> partitionColumnMetadata = tableMetadata
+                  .getRecordsByKeyPrefixes(
+                      HoodieListData.lazy(generateColumnStatsKeys(colsToIndex, partitionName)),
```

Comment:
as we discussed yesterday, we already have list of file names we are interested in. So, we should include those as well. currently, we are only looking up for col name and partition name

Reply:
I'm actually worried about the size of exact lookup keys given the combination of columns and files (e.g., 32 columns into 10K files at a large scale would result in 320K lookup keys).  Since this is not a necessary change for reducing the list calls, I haven't put it in.